ARCHITECTURE.md

```diff
@@ -0,0 +1,84 @@
+# Exonum Architecture
+
+This document describes the high-level architecture of Exonum. Its target
+audience is developers of Exonum core. To learn how to *use* Exonum, see
+[documentation].
+
+[documentation]: https://exonum.com/doc/
+
+This document is provided on the best effort basis — it is not guaranteed to be
+up-to-date and complete. The purpose of this document is to highlight the most
+important types, modules and functions in the code, to allow new contributors to
+dive into Exonum quickly. It is assumed that the reader is familiar with core
+concepts of Exonum, such as [services] and [transactions].
+
+[services]: https://exonum.com/doc/architecture/services/
+[transactions]: https://exonum.com/doc/architecture/transactions/
+
+## Code organization
+
+The repository is a single Cargo workspace of several related packages:
+
+- `exonum`, most of the code lives here,
+
+- `sandbox`, where integration tests live,
+
+- `testkit`, a framework for testing Exonum services.
+
+
+This document currently describes only the `exonum` crate.
+
+## Storage
+
+Storage module can be found at `src/storage`. It provides an abstraction over an
+embedded key value store: the `Database` trait. The keys and values are just
+slices of bytes: `&[u8]`.
+
+To transform raw bytes into native Rust data structures, the `StorageKey` and
+`StorageValue` traits are used.
+
+On top of raw storage, collections like sets, lists and maps are provided: see
+various `*Index` structs. Of particular interest are `storage/proof_map_index`
+and `storage/proof_list_index` modules, which provide Merkelized collections,
+capable of providing compact proofs for search queries.
+
+Note that indexes are used both by the Exonum core and by the services. The core
```

Comment:
Which services (the seems redundant, @Maria-Nosyk please correct me if I'm wrong)?

Reply:
I think "the services which were discussed in `the reader is familiar with core concepts of Exonum`", so `the` seems slightly more correct to me.